engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/AtomWithAHoleNode.java

```diff
@@ -0,0 +1,196 @@
+    if (result instanceof Atom atom) {
+      var index = swapNode.findHoleIndex(atom, lazy);
+      if (index >= 0) {
+        var function = swapNode.createFn(lazy);
+        lazy.init(atom, index, function);
+        return lazy;
+      }
+    }
+    throw new PanicException(ctx.getBuiltins().error().makeUninitializedStateError(result), this);
+  }
+
+  @ExportLibrary(InteropLibrary.class)
+  static final class HoleInAtom implements TruffleObject {
+    Atom result;
+    int index;
+    Function function;
+
+    HoleInAtom() {
+    }
+
+    void init(Atom result, int index, Function function) {
+      this.result = result;
+      this.index = index;
+      this.function = function;
+    }
+
+    @ExportMessage boolean hasMembers() {
+        return true;
+    }
+
+    @ExportMessage boolean isMemberReadable(String member) {
+        return switch (member) {
+            case "value", "fill" -> true;
+            default -> false;
+        };
+    }
+
+    @ExportMessage Object getMembers(boolean includeInternal) throws UnsupportedMessageException {
+        return new Array("value", "fill");
+    }
+
+    @ExportMessage
+    Object readMember(String name) throws UnknownIdentifierException {
+      if ("value".equals(name)) {
+        return result;
+      }
+      if ("fill".equals(name)) {
+        return function;
+      }
+      throw UnknownIdentifierException.create(name);
+    }
+
+    @ExportMessage
+    String toDisplayString(boolean pure) {
+      return "Meta.atom_with_hole";
+    }
+  }
+  static final class SwapAtomFieldNode extends RootNode {
+    private final FunctionSchema schema;
+    private final ValueProfile sameAtom = ValueProfile.createClassProfile();
+    @CompilerDirectives.CompilationFinal
+    private int lastIndex = -1;
+
+    private SwapAtomFieldNode() {
+      super(null);
+      this.schema = new FunctionSchema(FunctionSchema.CallerFrameAccess.NONE, new ArgumentDefinition[]{
+        new ArgumentDefinition(0, "lazy", ArgumentDefinition.ExecutionMode.EXECUTE),
+        new ArgumentDefinition(1, "value", ArgumentDefinition.ExecutionMode.EXECUTE)
+      }, new boolean[]{
+        true, false
+      }, new CallArgumentInfo[0]);
+    }
+
+    static SwapAtomFieldNode create() {
+      return new SwapAtomFieldNode();
+    }
+
+    int findHoleIndex(Atom atom, HoleInAtom lazy) {
+      var arr = atom.getFields();
+      if (lastIndex >= 0 && lastIndex < arr.length) {
+        if (arr[lastIndex] == lazy) {
+          return lastIndex;
+        }
+      }
+      int index = findHoleIndexLoop(arr, lazy);
+      if (index == -1) {
+        return -1;
+      }
+      if (lastIndex == -1) {
+        lastIndex = index;
+        CompilerDirectives.transferToInterpreterAndInvalidate();
+        return index;
+      } else {
+        if (lastIndex != -2) {
+          CompilerDirectives.transferToInterpreterAndInvalidate();
+          lastIndex = -2;
```

Comment:
it would be nice if those magic constants `-1` and `-2` were documented somehow.